### ISSUE TITLE
Run with libprotobuf >= 3.0 in travis

### ISSUE THREAD

**Issue by tomas-abrahamsson** [tomas-abrahamsson/gpb#144]
Spin-off from #142

**Comment by fenollp** [comment]
Pasting this here for now: https://github.com/tomas-abrahamsson/gpb/pull/142#issuecomment-405902384
I'll have a try with it later.

**Comment by fenollp** [comment]
Looks like I found a bug in Erlang: https://travis-ci.org/fenollp/gpb/jobs/406285648

```erlang
1> string:lexemes("\n\r\n", "\r").
["\n\r\n"]
2> string:lexemes("a\ra", "\r").
["a","a"]
```

**Comment by tomas-abrahamsson** [comment]
I experimented a bit, with (a) [a ppa I found for trusty](https://launchpad.net/~maarten-fonville/+archive/ubuntu/protobuf), and (b) building protobuf from source. In the table below, I also included a the 4.3.1 build for reference. (This is without running any quickcheck tests)

approach | commit | build link(s)  | protobuf version | avg time per build | total elapsed wallclock time(s)
--------------|-----------|------------------|------------------------|-------------------------|--------------------------------------
baseline for reference | 7455ecd (=4.3.1) | [137](https://travis-ci.org/tomas-abrahamsson/gpb/builds/422623745) | 2.5.0 | ~2m50s | ~20m
(a) use ppa | 509871e | [145](https://travis-ci.org/tomas-abrahamsson/gpb/builds/423455383) | 3.5.1 | ~3m30s | ~25m
(b) build protobuf | 9c995aaf | [148](https://travis-ci.org/tomas-abrahamsson/gpb/builds/423482626), [149](https://travis-ci.org/tomas-abrahamsson/gpb/builds/423489848) | 3.6.1 | ~15m | ~1h50m, ~1h40m

I considered adopting your docker approach, but I realized that not only must I use `protoc` from the docker container to generate the C++ code for the nif. At runtime I will also need to convince the dynamic loader to resolve symbols from the eunit's _xyz_.nif.so on the host and the libprotobuf.so inside the docker container (and any dependency it might have in turn). Interesting challenge. Or maybe link things statically, but that's not too tempting because then I wouldn't be testing what I think most nif user's run.

I think I'll be going with the ppa approach. Should it in some future begin to fail, I can always fall back to building from source.

**Comment by fenollp** [comment]
Nice investigation!
For a multi-language app & the sake of simplicity I am using this now:
```
docker run --rm znly/protoc:0.3.0 --version
libprotoc 3.5.1
```
But yeah YMMV.

**Comment by tomas-abrahamsson** [comment]
I see. For now, I went with the ppa and included it in the just-pushed 4.3.2, hence closing this.
